content/en/blog/_posts/2021-08-30-volume-populators-alpha.md

```diff
@@ -0,0 +1,207 @@
+VolumeSnapshots in the same namespace, and modified the normal provisioning process so that
+instead of yielding an empty volume, the new PVC contained the same data as either the cloned
+PVC or the cloned VolumeSnapshot.
+
+Volume populators embrace the same design idea, but extend it to any type of object, as long
+as there exists a CRD to define the data source and a populator controller to implement the
+logic. Initially the `dataSource` field was directly extended to allow arbitrary objects, if
+the `AnyVolumeDataSource` feature gate was enabled on the cluster. That change unfortunately
+caused backwards compatibility problems, and so the new `dataSourceRef` field was born.
+
+In v1.22 if the `AnyVolumeDataSource` feature gate is enabled, the `dataSourceRef` field is
+added, which behaves similarly to the `dataSource` field except that it allows arbitrary
+objects to be specified. The API server ensures that the two fields always have the same
+contents, and neither of them are mutable. The differences is that at creation time
+`dataSource` allows only PVCs or VolumeSnapshots, and ignores all other values, while
+`dataSourceRef` allows most types of objects, and in the few cases it doesn't allow an
+object (core objects other than PVCs) a validation error occurs.
+
+This API redesign will eventually result in a preference for the `dataSourceRef` field for
+all use cases, but in the immediate timeframe it is used by volume populators.
+
+## Using populators
+
+Every volume populator must have one or more CRDs that it supports. Administrators may
+install the CRD and the populator controller and then PVCs with a `dataSourceRef` specifies
+a CR of the type that the populator supports will be handled by the populator controller
+instead of the CSI plugin directly.
+
+Underneath the covers, the CSI plugin is still invoked to create an empty volume, which
+the populator controller fills with the appropriate data. The PVC doesn't bind to the PV
+until it's fully populated, so it's safe to define a whole application manifest including
+pod and PVC specs and the pods won't being running until everything is ready, just as if
+the PVC was a clone of another PVC or VolumeSnapshot.
+
+## How it works
+
+PVCs with data sources are still noticed by the external-provisioner sidecar for the
+related storage class (assuming a CSI provisioner is used), but because the sidecar
+doesn't understand the data source kind, it doesn't do anything. The populator controller
+is also watching for PVCs with data sources of a kind that it understands and when it
+sees one, it creates a temporary volume of the same size, volume mode, storage class,
+and even on the same topology (if topology is used) as the original volume. The populator
+controller creates a worker pod that attaches to the volume and writes the necessary
+data to it, then detaches from the volume and the populator controller rebinds the
+temporary volume to the orignal PVC.
+
+## Important components
```

Comment:
```suggestion
## Trying it out
```
(I _hope_ folks aren't going to jump into a production deploy whilst reading this&hellip;)